Update primary-constructors.md (#6806)
dotnet/csharplang: proposals/primary-constructors.md

>     - Otherwise, the result is the same as a member access ([§11.7.6](expressions.md#1176-member-access)) of the form `T.I` or `T.I<A₁, ..., Aₑ>`.
>   - **Otherwise, if the declaration of `T` includes a primary constructor parameter `I`, the result is the primary constructor parameter `I`.**

The first addition corresponds to the change incurred by [primary constructors on records](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/records.md#primary-constructor), and ensures that primary constructor parameters are found before any corresponding fields within initializers and base class arguments. It extends this rule to static initializers as well. However, since records always have an instance member with the same name as the parameter, the extension can only lead to a change in an error message. Illegal access to a parameter vs. illegal access to an instance member.  

The second addition allows primary constructor parameters to be found elsewhere within the type body, but only if not shadowed by members.

> It is an error to reference a primary constructor parameter if the reference does not occur within a `nameof` argument and
> at the same time does not occur within one of the following:

Reviewer: Could the `nameof` case be included in the cases below?
```suggestion
It is an error to reference a primary constructor parameter if the reference does not occur within one of the following:
- a `nameof` argument
```

> - an initializer of an instance field, property or event of the declaring type (type declaring primary constractor with the parameter).

Reviewer: ```suggestion
- an initializer of an instance field, property or event of the declaring type (type declaring primary constructor with the parameter).
```

- the `argument_list` of `class_base` of the declaring type.
- the body of an instance method (note that instance constructors are excluded) of the declaring type.
- the body of an instance accessor of the declaring type.

> In other words, primary constructor parameters are in scope throughout the declaring type body. They shadow declaring type members within

Reviewer 2: ```suggestion
In other words, primary constructor parameters are in scope throughout the declaring type body. They shadow members of the declaring type within
```

> an initializer of a field, property or event of the declaring type, or within the `argument_list` of `class_base` of the declaring type. They are shadowed by declaring type members everywhere else.

Reviewer 2: ```suggestion
an initializer of a field, property or event of the declaring type, or within the `argument_list` of `class_base` of the declaring type. They are shadowed by members of the declaring type everywhere else.
```

    
    // generated members
    private string __s; // for capture of s
    public C(bool b, int i, string s)
    {
        __s = s; // capture s
        I = i; // run I's initializer

> 	B(b) // run B's constructor

Reviewer 3: Nit: inconsistent spacing (1 tab vs 8 spaces on the previous line).

> 	B(b) // run B's constructor

Author: ```suggestion
        B(b) // run B's constructor
```

    }
}
```

It is an error for a non-primary constructor declaration to have the same parameter list as the primary constructor. All non-primary constructor declarations must use a `this` initializer, so that the primary constructor is ultimately called.

> Records produce a warning if primary constructor parameter isn't red within instance initializers or base initializer. Similar warning will be reported for primary constructor parameters in classes and structures:

Reviewer 3: ```suggestion
Records produce a warning if primary constructor parameter isn't read within instance initializers or base initializer. Similar warnings will be reported for primary constructor parameters in classes and structures:
```

> Records produce a warning if primary constructor parameter isn't read within instance initializers or base initializer. Similar warnings will be reported for primary constructor parameters in classes and structures:

Reviewer 2: ```suggestion
Records produce a warning if a primary constructor parameter isn't read within the (possibly generated) instance initializers or base initializer. Similar warnings will be reported for primary constructor parameters in classes and structures:
```

> - for a by-value parameter, if the parameter is not captured and is not red within instance initializers or base initializer.

Reviewer 3: ```suggestion
- for a by-value parameter, if the parameter is not captured and is not read within any instance initializers or base initializer.
```

> - for an `in` parameter, if the parameter is not red within instance initializers or base initializer. 

Reviewer 3: ```suggestion
- for an `in` parameter, if the parameter is not read within any instance initializers or base initializer. 
```

> - for a `ref` parameter, if the parameter is not red or written to within instance initializers or base initializer. 

Reviewer 3: ```suggestion
- for a `ref` parameter, if the parameter is not read or written to within any instance initializers or base initializer. 
```
